lightwallet-client-lib/src/main/java/co/electriccoin/lightwallet/client/internal/BlockingLightWalletClientImpl.kt

```diff
@@ -0,0 +1,191 @@
+                    .getLatestBlock(Service.ChainSpec.newBuilder().build())
+
+                val blockHeight = BlockHeightUnsafe(response.height)
+
+                Response.Success(blockHeight)
+            }
+        } catch (e: StatusRuntimeException) {
+            GrpcStatusResolver.resolveFailureFromStatus(e)
+        }
+    }
+
+    @Suppress("SwallowedException")
+    override fun getServerInfo(): Response<LightWalletEndpointInfoUnsafe> {
+        return try {
+            val lightdInfo = requireChannel().createStub(singleRequestTimeout)
+                .getLightdInfo(Service.Empty.newBuilder().build())
+
+            val lightwalletEndpointInfo = LightWalletEndpointInfoUnsafe.new(lightdInfo)
+
+            Response.Success(lightwalletEndpointInfo)
+        } catch (e: StatusRuntimeException) {
+            GrpcStatusResolver.resolveFailureFromStatus(e)
+        }
+    }
+
+    override fun submitTransaction(spendTransaction: ByteArray): Response<SendResponseUnsafe> {
+        if (spendTransaction.isEmpty()) {
+            return Response.Failure.Client.SubmitEmptyTransaction()
+        }
+        return try {
+            val request =
+                Service.RawTransaction.newBuilder().setData(ByteString.copyFrom(spendTransaction))
+                    .build()
+            val response = requireChannel().createStub().sendTransaction(request)
+
+            val sendResponse = SendResponseUnsafe.new(response)
+
+            Response.Success(sendResponse)
+        } catch (e: StatusRuntimeException) {
+            GrpcStatusResolver.resolveFailureFromStatus(e)
+        }
+    }
+
+    override fun fetchTransaction(txId: ByteArray): Response<RawTransactionUnsafe> {
+        if (txId.isEmpty()) {
+            return Response.Failure.Client.NullIdTransaction()
+        }
+        return try {
+            val request = Service.TxFilter.newBuilder().setHash(ByteString.copyFrom(txId)).build()
+
+            val response = requireChannel().createStub().getTransaction(request)
+
+            val transactionResponse = RawTransactionUnsafe.new(response)
+
+            Response.Success(transactionResponse)
+        } catch (e: StatusRuntimeException) {
+            GrpcStatusResolver.resolveFailureFromStatus(e)
+        }
+    }
+
+    override fun fetchUtxos(
+        tAddress: String,
+        startHeight: BlockHeightUnsafe
+    ): List<Service.GetAddressUtxosReply> {
+        val result = requireChannel().createStub().getAddressUtxos(
+            Service.GetAddressUtxosArg.newBuilder().setAddress(tAddress)
+                .setStartHeight(startHeight.value).build()
+        )
+        return result.addressUtxosList
+    }
+
+    override fun getTAddressTransactions(
+        tAddress: String,
+        blockHeightRange: ClosedRange<BlockHeightUnsafe>
+    ): Sequence<Service.RawTransaction> {
+        if (blockHeightRange.isEmpty() || tAddress.isBlank()) return emptySequence()
```

Comment:
Long-term, should we consider throwing an IllegalArgumentException if these arguments are not correct?

Reply:
Yep, understand. The proposed separation of logic sounds good to me. Stronger types are the right long-term solution.